src/io/github/sspanak/tt9/ime/modes/ModePredictive.java

```diff
@@ -86,18 +95,102 @@ public boolean onNumber(Language l, int key, boolean hold, boolean repeat) {
 	}
 
 
+	@Override
 	public void reset() {
 		super.reset();
 		digitSequence = "";
 		stem = "";
 	}
 
 
+	/**
+	 * shouldAddAutoSpace
+	 * When the "auto-space" settings is enabled, this determines whether to automatically add a space
+	 * at the end of a sentence or after accepting a suggestion. This allows faster typing, without
+	 * pressing space.
+	 *
+	 * See the helper functions for the list of rules.
+	 */
+	@Override
+	public boolean shouldAddAutoSpace(EditorInfo inputField, boolean isWordAcceptedManually, int incomingKey) {
+		return
+			shouldAddAutoSpaceAfterPunctuation(inputField, incomingKey)
+			|| shouldAddAutoSpaceAfterWord(inputField, isWordAcceptedManually, incomingKey);
+	}
+
+
+	/**
+	 * shouldDeletePrecedingSpace
+	 * When the "auto-space" settings is enabled, determine whether to delete spaces before punctuation.
+	 * This allows automatic conversion from: "words ." to: "words."
+	 */
+	@Override
+	public boolean shouldDeletePrecedingSpace(EditorInfo inputField) {
+		return
+			settings.getAutoSpace()
+			&& (
+				lastAcceptedWord.endsWith(".")
+				|| lastAcceptedWord.endsWith(",")
+				|| lastAcceptedWord.endsWith(";")
+				|| lastAcceptedWord.endsWith(":")
+				|| lastAcceptedWord.endsWith("!")
+				|| lastAcceptedWord.endsWith("?")
+				|| lastAcceptedWord.endsWith(")")
+				|| lastAcceptedWord.endsWith("]")
+				|| lastAcceptedWord.endsWith("'")
+			)
+			&& InputFieldHelper.isRegularTextField(inputField);
+	}
+
+
+	/**
+	 * shouldAddAutoSpaceAfterPunctuation
+	 * Determines whether to automatically adding a space after certain punctuation signs makes sense.
+	 * The rules are similar to the ones in the standard Android keyboard (with some exceptions,
+	 * because we are not using a QWERTY keyboard here).
+	 */
+	private boolean shouldAddAutoSpaceAfterPunctuation(EditorInfo inputField, int incomingKey) {
+		return
+			incomingKey != 0
+			&& settings.getAutoSpace()
+			&& (
+				lastAcceptedWord.endsWith(".")
+				|| lastAcceptedWord.endsWith(",")
+				|| lastAcceptedWord.endsWith(";")
+				|| lastAcceptedWord.endsWith(":")
+				|| lastAcceptedWord.endsWith("!")
+				|| lastAcceptedWord.endsWith("?")
+				|| lastAcceptedWord.endsWith(")")
+				|| lastAcceptedWord.endsWith("]")
+			)
+			&& InputFieldHelper.isRegularTextField(inputField);
+	}
+
+
+	/**
+	 * shouldAddAutoSpaceAfterPunctuation
+	 * Similar to "shouldAddAutoSpaceAfterPunctuation()", but determines whether to add a space after
+	 * words.
+	 */
+	private boolean shouldAddAutoSpaceAfterWord(EditorInfo inputField, boolean isWordAcceptedManually, int incomingKey) {
+		return
+			// 0 and 1 do not type letters. They are handled in "shouldAddAutoSpaceAfterPunctuation()".
+			incomingKey != 0 && incomingKey != 1
+			// Do not add space when auto-accepting words, because it feels very confusing when typing.
+			&& isWordAcceptedManually
+			// Emoji
+			&& !lastAcceptedSequence.startsWith("11")
```

Comment:
```suggestion
			&& !lastAcceptedSequence.startsWith("1")
```

Otherwise, it allows a space after punctuation.

Reply:
Done.